### DISCUSSION TITLE
DataSet.rotate and DataSet.translate execution times - path finder algorithm

### DISCUSSION THREAD

**Discussion by luizenger** [pyvista/pyvista#4147] [Q&A]
Hello there,

I am using PyVista to build a digital twin of a robot. Blocks are created mainly as pv.Cube, some merged together, some not. To make it move around I use the DataSet.translate() and DataSet.rotate_vector() methods. With inplace=False I check if the new mesh position causes a collision and update the original mesh if it doesn't.
The next step is to implement a path finder algorithm, so the robot can get to a distance target position following a path that won't cause any collision.
Initially I thought of implementing the less complex A* algorithm, but before that I tested how long it takes to detect a collision.
And it is actually way too slow to use such path finder, where several movements/positions have to be tested.

Surprisingly to me, what is slow is not detecting collision between meshes, but running DataSet.translate( ) and DataSet.rotate_vector( ) itself!
With a single and very simple pv.Cube mesh only, 1000 translations took me 300 ms, while 1000 rotations took 460 ms.

Is there a way to speed this up? Any other suggestions are appreciated!

Thank you.

**Comment by banesullivan** [comment]
That is a bit surprising... could you share some sort of minimal example to reproduce this slow translation?

**Reply by luizenger** [reply]
Sure:

```
import pyvista
import random

def speed_test(n):
    random.seed(1)

    cube = pyvista.Cube()

    for _ in range(n):
        rotvec = (random.random(), random.random(), random.random())
        # cube = cube.rotate_vector(rotvec, random.random(), (0, 1, 1))
        cube = cube.translate(rotvec)

n = 1000

%timeit -r 10 -n 1 speed_test(n)
```

Edit: properly formatting code

**Reply by banesullivan** [reply]
Results for me:

```
127 ms ± 1.09 ms per loop (mean ± std. dev. of 10 runs, 1 loop each)
```

-----

I wanted to compare to some other scenarios: The above code is as fast or faster than instantiating `n`-many cubes:

```py
def base_test(n):
    for _ in range(n):
        pyvista.Cube()
    
%timeit -r 10 -n 1 base_test(n)
```

```
174 ms ± 3.72 ms per loop (mean ± std. dev. of 10 runs, 1 loop each)
```

**Reply by adeak** [reply]
I've run a modified version of the code through cProfile and [snakeviz](https://jiffyclub.github.io/snakeviz/):
```py
import pyvista
import random

def speed_test(n):
    random.seed(1)

    cube = pyvista.Cube()

    for _ in range(n):
        rotvec = (random.random(), random.random(), random.random())
        # cube = cube.rotate_vector(rotvec, random.random(), (0, 1, 1))
        cube = cube.translate(rotvec)

n = 100_000
speed_test(n)
```
I increased `n` because otherwise most of the time is spent with importing the library and other fluff like that.

The result looks something like this:
![Screenshot from 2023-03-21 00-06-37](https://user-images.githubusercontent.com/17914410/226484943-e2c8c1e0-e6df-4883-829a-982f28ce3933.png)

90% of the time is indeed spent in `DataSetFilters.transform()`. One level lower we have
 * `_get_output()` of `pyvista/core/filters/` for 27% of the total time (7% of which is `pyvista.wrap()`)
 * `_update_alg()` in the same place for 9% of the total time
 * `DataSet.copy_from()` for 20% of the total time (but out of the 48% of `_get_output()` + `copy_from()`, 34% comes from `DataSet.copy_meta_from()` of which 21% is `copy.deepcopy()`)
 * `DataSet.points` for 7% of the total time
 * `DataSetAttributes.__getitem__()` -> `DataSetAttributes.get_array()` for 7% of the total time
 * lots of other, smaller fish
 
(All of the above are meant as percentage points relative to total runtime rather than successive percentage levels.)

Not sure we'll be able to do with most of these, but perhaps worth looking at sometime to see if there's some bottleneck we can address.

**Reply by luizenger** [reply]
Thanks for giving this a deeper look. I guess I'll work on finding another way to implement a path finder algorithm. The "digital twin" of the robot built using PyVista is really handy to see how a programmed movement will actually be executed and to stop at any possible collision, so I want to keep using it for the moment.

**Reply by akaszynski** [reply]
@luizenger, VTK filters generally have a bit of overhead as they've been tested and patched for a variety of issues over the years. This makes them quite robust, but as in the case you've discovered, sometimes slow.

I'd recommend rotating and translating by directly acting on the [points](https://docs.pyvista.org/api/core/_autosummary/pyvista.DataSet.points.html#points) attribute in situ using `+=`.

As for rotation about a vector, this should work:

```py
import random
import numpy as np
from scipy.spatial.transform import Rotation
import pyvista as pv


def rotate_points(points, axis, angle_degrees):
    # Normalize the rotation axis
    axis_normalized = axis / np.linalg.norm(axis)

    # Convert the angle from degrees to radians
    angle_radians = np.radians(angle_degrees)

    # Create a Rotation object using the axis-angle representation
    rotation = Rotation.from_rotvec(axis_normalized * angle_radians)

    # Rotate the points
    rotated_points = rotation.apply(points)

    return rotated_points

# Example usage:
mesh = pv.Sphere()

axis = np.random.random(3)
angle_degrees = random.random()*20

# rotate using pyvista (this creates a new data object
rotated_pv = mesh.rotate_vector(axis, angle_degrees)

mesh.points[:] = rotate_points(mesh.points, axis, angle_degrees)

assert np.allclose(mesh.points, rotated_pv.points)
```

This example leverages our [existing example](https://docs.pyvista.org/api/core/_autosummary/pyvista.DataSet.rotate_vector.html#rotate-vector) as a basis and expands on it by providing a simple function to rotate points about an axis.

With zero optimizations, this is about 5x faster:

```py
>>> timeit mesh.rotate_vector(axis, angle_degrees)
342 µs ± 3.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

>>> timeit rotate_points(mesh.points, axis, angle_degrees)
56.8 µs ± 80.4 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

**Reply by adeak** [reply]
Depending on how complex the movements are, I'd also consider separating the collision logic from the presentation (creating actual cubes). Working with only arrays of 8 corner points or some other useful representation of pieces would be a lot faster. I.e. not doing this part with PyVista, only visualization. Or if most of the non-collision cases also have no overlapping bounding boxes, keep track of the bounding box and check that first for a chance of collision. Stuff like that.

**Reply by luizenger** [reply]
Thank you @akaszynski and sorry for taking a long time to reply. I'll try that! 
And yes @adeak I totally agree that performing the collision logic with only corner points would be easier. I actually am using the merge filter to create more elaborate meshes with different Cubes. I could extract the corner points after merging and use those in collision detection. The first bottleneck I've encountered was the moving around itself, I can do that later.

**Comment by luizenger** [comment]
Which is more or less expected, isn't it? The translate method actually
creates a new mesh then copies it into the calling mesh if inplace=True. At
least that was my understanding when taking a quick look into the source
code.

Bane Sullivan ***@***.***> schrieb am Mo., 20. März 2023,
15:54:
